service/module.go

```diff
@@ -23,10 +23,19 @@ package service
 import (
 	"fmt"
 	"sync"
-
-	"github.com/uber-go/tally"
 )
 
+// ModuleProvider provides Modules.
+type ModuleProvider interface {
+	// Name is the name that will be used for a new Module
+	// if no name is given as a ModuleOption.
+	Name() string
+	// Create a new Module. The name of the Host and the scoping
+	// of associated functions on the Host will be done using a name
+	// provided by a ModuleOption, or by the Name on this ModuleProvider.
+	Create(Host) (Module, error)
```

Comment:
It can be part of the interface but still made simple to implement:

```go
type moduleProviderFunc struct {
  name string
  fn moduleCreateFunc
}

func (m moduleProviderFunc) Name() string { return m.name }
func (m moduleProviderFunc) Create(Host) (Module, error) {
  return m.fn(host)
}

func ModuleProviderFromFunc(f moduleCreateFunc, n name) ModuleProvider {
  return moduleProviderFunc{name: n, fn: f}
}
```